File: 02_Data_Structures/trees/b_trees/implementations/operations.cpp
// File: operations.cpp
// Topic: B-Tree Operations (Insert, Search, Traverse)
//
// Description:
// This file implements key operations on a B-Tree:
// 1. Search
// 2. Insert
// 3. Node split
// 4. Traversal
//
// B-Tree Properties:
// - Multi-way search tree optimized for disk access
// - Every node can have multiple keys
// - All leaves at same depth
// - Keys inside node are sorted
//
// Complexity:
// Search: O(log n)
// Insert: O(log n)
// Traverse: O(n)
//
// Note: Implementation corresponds to a B-Tree of minimum degree t.
//       Each node contains between t-1 to 2t-1 keys.
//

#include <iostream>
using namespace std;

// -------------------- B-Tree Node --------------------
class BTreeNode {
public:
    int *keys;          // array of keys
    int t;              // minimum degree
    BTreeNode **child;  // array of child pointers
    int n;              // current number of keys
    bool leaf;          // is leaf?

    BTreeNode(int t, bool leaf);

    void traverse();               // print all keys
    BTreeNode* search(int k);      // search key
    void insertNonFull(int k);     // insert key in non-full node
    void splitChild(int i, BTreeNode *y); // split child
};

// -------------------- B-Tree --------------------
class BTree {
public:
    BTreeNode *root;
    int t;

    BTree(int t) {
        root = nullptr;
        this->t = t;
    }

    void traverse() {
        if (root) root->traverse();
    }

    BTreeNode* search(int k) {
        return (root == nullptr) ? nullptr : root->search(k);
    }

    void insert(int k);
};

// --------------------------------------------------------
// Node constructor
// --------------------------------------------------------
BTreeNode::BTreeNode(int t, bool leaf) {
    this->t = t;
    this->leaf = leaf;

    keys = new int[2 * t - 1];
    child = new BTreeNode*[2 * t];

    n = 0;
}

// --------------------------------------------------------
// Traversal (Inorder-like within nodes)
// --------------------------------------------------------
void BTreeNode::traverse() {
    for (int i = 0; i < n; i++) {
        if (!leaf)
            child[i]->traverse();
        cout << keys[i] << " ";
    }

    if (!leaf)
        child[n]->traverse();
}

// --------------------------------------------------------
// Search key in subtree rooted at this node
// --------------------------------------------------------
BTreeNode* BTreeNode::search(int k) {
    int i = 0;

    // find correct key position
    while (i < n && k > keys[i])
        i++;

    // if key found
    if (i < n && keys[i] == k)
        return this;

    if (leaf)
        return nullptr;

    return child[i]->search(k);
}

// --------------------------------------------------------
// Split child y at index i
// --------------------------------------------------------
void BTreeNode::splitChild(int i, BTreeNode *y) {
    BTreeNode *z = new BTreeNode(y->t, y->leaf);

    z->n = t - 1;

    // transfer last t-1 keys
    for (int j = 0; j < t - 1; j++)
        z->keys[j] = y->keys[j + t];

    // transfer children if internal node
    if (!y->leaf) {
        for (int j = 0; j < t; j++)
            z->child[j] = y->child[j + t];
    }

    y->n = t - 1;

    // shift children of current node
    for (int j = n; j >= i + 1; j--)
        child[j + 1] = child[j];

    child[i + 1] = z;

    // shift keys
    for (int j = n - 1; j >= i; j--)
        keys[j + 1] = keys[j];

    keys[i] = y->keys[t - 1];
    n++;
}

// --------------------------------------------------------
// Insert key into non-full node
// --------------------------------------------------------
void BTreeNode::insertNonFull(int k) {
    int i = n - 1;

    if (leaf) {
        // find location to insert new key
        while (i >= 0 && keys[i] > k) {
            keys[i + 1] = keys[i];
            i--;
        }

        keys[i + 1] = k;
        n++;
    } else {
        // find child to insert key
        while (i >= 0 && keys[i] > k)
            i--;

        i++;

        // if child is full, split
        if (child[i]->n == (2 * t - 1)) {
            splitChild(i, child[i]);

            if (keys[i] < k)
                i++;
        }

        child[i]->insertNonFull(k);
    }
}

// --------------------------------------------------------
// Insert key in B-tree
// --------------------------------------------------------
void BTree::insert(int k) {
    if (root == nullptr) {
        root = new BTreeNode(t, true);
        root->keys[0] = k;
        root->n = 1;
        return;
    }

    if (root->n == (2 * t - 1)) {
        BTreeNode *s = new BTreeNode(t, false);
        s->child[0] = root;

        s->splitChild(0, root);

        int i = 0;
        if (s->keys[0] < k)
            i++;

        s->child[i]->insertNonFull(k);
        root = s;
    } else {
        root->insertNonFull(k);
    }
}

// --------------------------------------------------------
// Example usage (remove in production)
// --------------------------------------------------------
/*
int main() {
    BTree t(3); // degree

    t.insert(10);
    t.insert(20);
    t.insert(5);
    t.insert(6);
    t.insert(12);
    t.insert(30);
    t.insert(7);
    t.insert(17);

    cout << "Traversal: ";
    t.traverse();
    cout << endl;

    int key = 6;
    cout << "Search " << key << ": ";
    cout << (t.search(key) ? "Found" : "Not Found") << endl;
}
*/

File: 02_Data_Structures/trees/b_trees/implementations/operations.py
# File: operations.py
# Topic: B-Tree Operations (Insert, Search, Traverse)
#
# Problem Statement:
# Implement a B-Tree (of minimum degree t) supporting:
#   1) search(key) -> bool
#   2) insert(key)
#   3) traverse() -> in-order-like print/iterate of keys
#
# Notes:
# - B-Tree is a multi-way balanced search tree optimized for block I/O.
# - Each node holds between (t-1) and (2t-1) keys (root is a special case).
# - All leaves are at the same depth.
# - Keys in a node are sorted; children pointers separate key ranges.
#
# Algorithm Choices:
# - Insertion uses top-down splitting: if a child is full, split it before descending.
# - Search is a standard node-local binary-like scan followed by descent.
# - Traversal prints keys in sorted order by recursively traversing children.
#
# Complexity:
# - Search:  O(log n)
# - Insert:  O(log n)
# - Space:   O(n)
#
# Conventions / Edge Cases:
# - Duplicate keys: this implementation IGNOREs duplicates (no multiset).
#   Change the condition in insert_non_full to allow duplicates to go to a side consistently.
# - Minimum degree t must be >= 2.

from __future__ import annotations
from typing import List, Optional, Iterable


class BTreeNode:
    """A B-Tree node."""

    def __init__(self, t: int, leaf: bool):
        assert t >= 2, "Minimum degree t must be >= 2"
        self.t = t
        self.leaf = leaf
        self.keys: List[int] = []            # sorted keys (len in [t-1, 2t-1], except root/leaves during growth)
        self.child: List[BTreeNode] = []     # children pointers (len = len(keys)+1 for internal nodes)

    # ---------- Traversal ----------
    def traverse(self) -> Iterable[int]:
        """Yield all keys in sorted order."""
        if self.leaf:
            for k in self.keys:
                yield k
        else:
            for i, k in enumerate(self.keys):
                # traverse left child i
                yield from self.child[i].traverse()
                # then yield key i
                yield k
            # final rightmost child
            yield from self.child[len(self.keys)].traverse()

    # ---------- Search ----------
    def search(self, k: int) -> Optional["BTreeNode"]:
        """Return the node containing key k, or None."""
        # find first index i s.t. keys[i] >= k (linear scan; could be binary search for micro-optimization)
        i = 0
        while i < len(self.keys) and k > self.keys[i]:
            i += 1

        if i < len(self.keys) and self.keys[i] == k:
            return self

        if self.leaf:
            return None

        return self.child[i].search(k)

    # ---------- Split ----------
    def split_child(self, i: int) -> None:
        """
        Split full child y = child[i] (which must have 2t-1 keys) into two nodes y,z
        and move median up into self.
        """
        t = self.t
        y = self.child[i]
        assert len(y.keys) == 2 * t - 1, "split_child called on non-full child"

        # New node z will store y's last t-1 keys (and children if internal)
        z = BTreeNode(t, y.leaf)
        # Keys: [0..t-2], [t-1], [t..2t-2]
        median_key = y.keys[t - 1]
        z.keys = y.keys[t:]           # last t-1
        y.keys = y.keys[:t - 1]       # first t-1

        if not y.leaf:
            z.child = y.child[t:]     # last t children
            y.child = y.child[:t]     # first t children

        # Insert z as a new child to the right of y
        self.child.insert(i + 1, z)
        # Move median key up into current node
        self.keys.insert(i, median_key)

    # ---------- Insert into Non-Full Node ----------
    def insert_non_full(self, k: int) -> None:
        """Insert key k into this node which is known to be non-full."""
        i = len(self.keys) - 1

        if self.leaf:
            # Place k into the correct position in keys (prevent duplicates)
            # Shift larger keys right, then insert
            # If you want to allow duplicates, change the strict '>' to '>=' or pick a side.
            while i >= 0 and self.keys[i] > k:
                i -= 1
            if i >= 0 and self.keys[i] == k:
                return  # ignore duplicates
            self.keys.insert(i + 1, k)
        else:
            # Find child index to descend
            while i >= 0 and self.keys[i] > k:
                i -= 1
            i += 1

            # If the child is full, split first
            if len(self.child[i].keys) == 2 * self.t - 1:
                self.split_child(i)
                # After split, decide which of the two children to descend into
                if k > self.keys[i]:
                    i += 1
                elif k == self.keys[i]:
                    return  # ignore duplicates

            self.child[i].insert_non_full(k)


class BTree:
    """B-Tree wrapper."""

    def __init__(self, t: int):
        assert t >= 2, "Minimum degree t must be >= 2"
        self.t = t
        self.root: Optional[BTreeNode] = None

    # ---------- Public API ----------
    def search(self, k: int) -> bool:
        """Return True if key k is present."""
        if self.root is None:
            return False
        return self.root.search(k) is not None

    def insert(self, k: int) -> None:
        """Insert key k into the B-Tree (ignores duplicates)."""
        if self.root is None:
            self.root = BTreeNode(self.t, True)
            self.root.keys = [k]
            return

        # If root is full, grow in height
        if len(self.root.keys) == 2 * self.t - 1:
            s = BTreeNode(self.t, False)
            s.child = [self.root]
            s.split_child(0)
            # select child to descend
            i = 0
            if k > s.keys[0]:
                i = 1
            elif k == s.keys[0]:
                self.root = s  # even if duplicate, root changed
                return

            s.child[i].insert_non_full(k)
            self.root = s
        else:
            self.root.insert_non_full(k)

    def traverse(self) -> Iterable[int]:
        """Yield keys in sorted order."""
        if self.root is None:
            return iter(())
        return self.root.traverse()


# ------------------ Example / Quick Test ------------------
# if __name__ == "__main__":
#     bt = BTree(t=3)  # min degree
#     data = [10, 20, 5, 6, 12, 30, 7, 17, 3, 2, 25]
#     for x in data:
#         bt.insert(x)
#
#     print("Traverse:", *bt.traverse())
#     for q in [6, 15, 25, 100]:
#         print(f"search({q}) ->", bt.search(q))
#
# Expected:
# Traverse: 2 3 5 6 7 10 12 17 20 25 30
# search(6) -> True
# search(15) -> False
# search(25) -> True
# search(100) -> False

